### PR TITLE
[MXNET-876] make CachedOp a normal operator

### PR DESCRIPTION
**Pull request by zheng-da**
## Description ##
Currently, CachedOp is used to execute the graph in a Gluon hybrid block when the block is hybridized. It's registered as an operator, but it doesn't have a full set of operator attributes. So it can't be used as a regular operator and can't be used in a normal NNVM computation graph. This PR is to extend CachedOp and make it a normal operator. The main motivation is to use it as a default subgraph operator, as proposed in [Unified integration with external acceleration libraries](https://cwiki.apache.org/confluence/display/MXNET/Unified+integration+with+external+acceleration+libraries).

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: include/mxnet/c_api.h [hunk 1]
```diff
@@ -1073,6 +1073,9 @@ MXNET_DLL int MXSymbolGetInputSymbols(SymbolHandle sym, SymbolHandle **inputs,
 MXNET_DLL int MXSymbolCutSubgraph(SymbolHandle sym, SymbolHandle **inputs,
                                   int *input_size);
 
+int MXMakeSubgraph(SymbolHandle sym, SymbolHandle *input_symbols, mx_uint num_inputs,
```
Comment (larroy) [inline include/mxnet/c_api.h]: Doc missing

File: python/mxnet/symbol/contrib.py [hunk 2]
```diff
@@ -336,3 +336,10 @@ def check_data(inputs, in_type, msg):
         states = states[0]
 
     return (outs, states)
+
+def make_subgraph(subg, *args):
```
Comment (larroy) [inline python/mxnet/symbol/contrib.py]: Doc missing

File: src/imperative/cached_op.cc [hunk 3]
```diff
@@ -1047,6 +1047,105 @@ void CachedOp::Backward(
   Engine::Get()->set_bulk_size(prev_bulk_size);
 }
 
+struct CachedOpActualState {
+  std::shared_ptr<CachedOp> op;
+  OpStatePtr forward_state;
+
+  CachedOpActualState(std::shared_ptr<CachedOp> op) {
```
Comment (larroy) [inline src/imperative/cached_op.cc]: By reference?

File: src/imperative/cached_op.cc [hunk 4]
```diff
@@ -1047,6 +1047,105 @@ void CachedOp::Backward(
   Engine::Get()->set_bulk_size(prev_bulk_size);
 }
 
+struct CachedOpActualState {
```
Comment (larroy) [inline src/imperative/cached_op.cc]: Missing class doc

File: src/imperative/cached_op.cc [hunk 5]
```diff
@@ -1047,6 +1047,105 @@ void CachedOp::Backward(
...
+  std::shared_ptr<CachedOp> op;
+  OpStatePtr forward_state;
+
+  CachedOpActualState(std::shared_ptr<CachedOp> op) {
+    this->op = op;
+  }
+};
+
+void CachedOpForward(const OpStatePtr& state_ptr,
```
Comment (larroy) [inline src/imperative/cached_op.cc]: Missing a short documentation stating what's the intention and how it works

File: src/imperative/cached_op.cc [hunk 6]
```diff
@@ -1047,6 +1047,105 @@ void CachedOp::Backward(
...
+  auto out_end = in_bufs.end();
+
+  for (auto it = ograds_begin; it != ograds_end; it++)
+    in_ptrs.push_back(&(*it));
+
+  const std::vector<bool> &save_inputs = s.op->save_inputs();
+  const std::vector<bool> &save_outputs = s.op->save_outputs();
+  CHECK_EQ(save_inputs.size(), in_end - in_begin);
+  CHECK_EQ(s.op->num_outputs(), out_end - out_begin);
+  for (auto it = in_begin; it != in_end; it++) {
```
Comment (larroy) [inline src/imperative/cached_op.cc]: ++it is potentially faster

File: src/imperative/cached_op.h [hunk 7]
```diff
@@ -116,6 +116,24 @@ class CachedOp {
       DispatchMode* dispatch_mode,
       std::vector<int> *in_attrs,
       std::vector<int> *out_attrs);
+  bool ForwardInferShape(
```
Comment (larroy) [inline src/imperative/cached_op.h]: Missing documentation in function prototypes

File: tests/python/unittest/test_operator.py [hunk 8]
```diff
@@ -6968,6 +6968,39 @@ def test_roi_align_autograd(sampling_ratio=0):
...
+    b_arr = mx.nd.random.uniform(-1, 1, shape=(10, 10), ctx=default_context())
+    e1 = d.bind(ctx=default_context(), args={'a': a_arr, 'b': b_arr},
+            args_grad={'a': mx.nd.empty(shape=(10, 10)),
+                       'b': mx.nd.empty(shape=(10, 10))})
+    e2 = y.bind(ctx=default_context(), args={'a1': a_arr, 'b1': b_arr},
+            args_grad={'a': mx.nd.empty(shape=(10, 10)),
+                       'b': mx.nd.empty(shape=(10, 10))})
+    e1.forward()
+    e2.forward()
+    assert_almost_equal(e1.outputs[0].asnumpy(), e2.outputs[0].asnumpy(),
```
Comment (larroy) [inline tests/python/unittest/test_operator.py]: Why almost equal and not equal?

**Comment by zheng-da** [comment]
@larroy thanks for the review. The code, especially the API design, is still experimental. I'll let you know when the code is ready for review.

File: tests/python/unittest/test_operator.py (hunk 8 above)
Comment (marcoabreu) [reply tests/python/unittest/test_operator.py]: I think due to floating point precision

**Comment by reminisce** [comment]
I think the function like shape, type inferences, FMutate, etc. used in operator registration should not belong to CachedOp only. They should be made generally available to subgraph-type operators, while CachedOp is just a special case.

File: src/c_api/c_api_symbolic.cc [hunk 9]
```diff
@@ -366,6 +367,36 @@ int MXSymbolGetInputSymbols(SymbolHandle sym, SymbolHandle **input_arr, int *inp
...
+  nnvm::NodeEntry entry = nnvm::MakeNode("_CachedOp", "_CachedOp", inputs);
+  nnvm::NodePtr n = entry.node;
+
+  // Create CachedOp for the node.
+  std::vector<std::pair<std::string, std::string> > kwargs;
+  kwargs.push_back(std::pair<std::string, std::string>("inline_limit", "0"));
+  n->attrs.parsed = std::make_shared<mxnet::CachedOp>(*s, kwargs);
+
+  // Create a new symbol for this node.
+  s = new nnvm::Symbol();
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply src/c_api/c_api_symbolic.cc]: i just follow the implementations in other APIs. The symbol will be saved in a Python symbol handle. AFAIK, once the python symbol handle is destroyed, the symbol object will also be destroyed.

File: src/imperative/cached_op.cc (hunk 6 above)
Comment (zheng-da) [reply src/imperative/cached_op.cc]: really? Where is this documented?

**Comment by zheng-da** [comment]
This PR should be rebased to https://github.com/apache/incubator-mxnet/pull/12157

File: src/operator/subgraph/common.h [hunk 10]
```diff
@@ -94,10 +93,15 @@ inline bool DefaultSubgraphOpShape(const nnvm::NodeAttrs& attrs,
...
 }
 
-inline bool DefaultSubgraphOpType(const nnvm::NodeAttrs& attrs,
-                                  std::vector<int> *in_types,
-                                  std::vector<int> *out_types) {
-  const nnvm::Symbol& subgraph_sym = *attrs.subgraphs[0];
+inline bool DefaultSubgraphOpShape(const nnvm::NodeAttrs& attrs,
+                                   std::vector<TShape> *in_shapes,
+                                   std::vector<TShape> *out_shapes) {
+  return DefaultSubgraphOpShape1(*attrs.subgraphs[0], in_shapes, out_shapes);
```
Comment (reminisce) [inline src/operator/subgraph/common.h]: Maybe rename `DefaultSubgraphOpShape1` to something like a helper function for better readability?

File: src/imperative/cached_op.cc [hunk 11]
```diff
@@ -1067,34 +1067,140 @@ void CachedOp::Backward(
...
-                                  const int dev_mask,
-                                  DispatchMode* dispatch_mode,
-                                  std::vector<int> *in_attrs,
-                                  std::vector<int> *out_attrs) {
-  using namespace imperative;
-  nnvm::Graph g(fwd_graph_);
-  const auto& idx = g.indexed_graph();
-  const auto &outputs = idx.outputs();
+/*
+ * This is the operator state of CachedOp when CachedOp is used in the symbol
```
Comment (reminisce) [inline src/imperative/cached_op.cc]: Please elaborate on the necessity of adding this data structure in the description.

File: src/imperative/cached_op.cc [hunk 12]
```diff
@@ -1067,34 +1067,145 @@ void CachedOp::Backward(
...
+  else
+    orig_is_train = Imperative::Get()->is_training();
+  // TODO(zhengda) is it right to use false here?
+  s.op->Backward(false, s.forward_state, in_ptrs, req, out_ptrs);
+  Imperative::Get()->set_is_training(orig_is_train);
+
+  // Clean up what we recorded.
+  s.forward_state.reset();
+
+  // The arrays in out_ptrs may be changed by CachedOp.
```
Comment (eric-haibin-lin) [inline src/imperative/cached_op.cc]: why would it be changed?

File: src/imperative/cached_op.cc [hunk 13]
```diff
@@ -1067,34 +1067,145 @@ void CachedOp::Backward(
...
+    out_ptrs.push_back(&out_bufs[i]);
+  CHECK_EQ(out_ptrs.size(), s.op->num_backward_outputs());
+  // Set is_training correct for the imperative executor.
+  bool orig_is_train;
+  if (ctx.is_train)
+    orig_is_train = Imperative::Get()->set_is_training(true);
+  else
+    orig_is_train = Imperative::Get()->is_training();
+  // TODO(zhengda) is it right to use false here?
+  s.op->Backward(false, s.forward_state, in_ptrs, req, out_ptrs);
```
Comment (eric-haibin-lin) [inline src/imperative/cached_op.cc]: please add more comment on `retain_graph=False`

File: src/imperative/cached_op.cc (hunk 12 above)
Comment (eric-haibin-lin) [reply src/imperative/cached_op.cc]: Thanks for updating the comments

**Pull request merged**